aten/src/ATen/native/cpu/group_norm_kernel.cpp

```diff
@@ -461,114 +558,285 @@ void ComputeInternalGradients(
+      const T_ACC* ds_ptr = ds + i * D;
+      const T_ACC* db_ptr = db + i * D;
       const int64_t g = i % G;
+
       for (const auto j : c10::irange(start, end)) {
         const int64_t c = g * D + j;
-        dgamma[c] += (ds_ptr[j] - db_ptr[j] * mean[i]) * rstd[i];
+        dgamma[c] += (ds_ptr[j] - db_ptr[j] * T_ACC(mean[i])) * T_ACC(rstd[i]);
       }
     }
   });
 }
 
-template <typename T>
-void BetaBackward(int64_t N, int64_t C, const T* db, T* dbeta) {
-  constexpr int64_t K = vec::Vectorized<T>::size();
+template <>
+void GammaBackward(
+    int64_t N,
+    int64_t C,
+    int64_t group,
+    const BFloat16* mean,
+    const BFloat16* rstd,
+    const float* ds,
+    const float* db,
+    BFloat16* dgamma) {
+  using fVec = vec::Vectorized<float>;
+  using bVec = vec::Vectorized<BFloat16>;
+  const int64_t G = group;
+  const int64_t D = C / G;
+  constexpr int64_t K = bVec::size();
+  float* acc_buffer = new float[C];
+  at::parallel_for(0, D, K, [=](int64_t start, int64_t end) {
+    for (const auto i : c10::irange(G)) {
+      std::memset(acc_buffer + i * D + start, 0, (end - start) * sizeof(float));
+    }
+    for (int64_t i = 0; i < N * G; ++i) {
+      const float* ds_ptr = ds + i * D;
+      const float* db_ptr = db + i * D;
+      const int64_t g = i % G;
+      for (const auto j : c10::irange(start, end)) {
+        const int64_t c = g * D + j;
+        acc_buffer[c] += (ds_ptr[j] - db_ptr[j] * float(mean[i])) * float(rstd[i]);
+      }
+    }
+  });
+
+  int64_t d = 0;
+  for (; d < C / K * K; d += K) {
+    fVec data_fvec0 = fVec::loadu(acc_buffer + d);
+    fVec data_fvec1 = fVec::loadu(acc_buffer + d + K / 2);
+    convert_float_bfloat16(data_fvec0, data_fvec1).store(dgamma + d);
+
+  }
+  if (C - d > 0) {
+    fVec data_fvec0 = fVec::loadu(acc_buffer + d, (C - d) > fVec::size() ? fVec::size() : (C - d));
+    fVec data_fvec1 = fVec::loadu(acc_buffer + d + fVec::size(), (C - d) > fVec::size() ? (C - d - fVec::size()) : 0);
+    convert_float_bfloat16(data_fvec0, data_fvec1).store(dgamma + d, C - d);
+  }
+  if (acc_buffer != nullptr) {
+    delete[] acc_buffer;
+  }
+}
+
+template <typename PT, typename T_ACC>
+void BetaBackward(int64_t N, int64_t C, const T_ACC* db, PT* dbeta) {
+  constexpr int64_t K = vec::Vectorized<PT>::size();
   at::parallel_for(0, C, K, [=](int64_t start, int64_t end) {
-    std::memset(dbeta + start, 0, (end - start) * sizeof(T));
+    std::memset(dbeta + start, 0, (end - start) * sizeof(PT));
     for (const auto i : c10::irange(N)) {
-      const T* db_ptr = db + i * C;
+      const T_ACC* db_ptr = db + i * C;
       for (const auto j : c10::irange(start, end)) {
         dbeta[j] += db_ptr[j];
       }
     }
   });
 }
 
-template <typename T>
+template <>
+void BetaBackward(int64_t N, int64_t C, const float* db, BFloat16* dbeta) {
+  using fVec = vec::Vectorized<float>;
+  using bVec = vec::Vectorized<BFloat16>;
+  constexpr int64_t K = bVec::size();
+  float* acc_buffer = new float[C];
```

Comment:
ditto

Reply:
Remove the specialized version for PT = BFloat16 since parameters should not be bfloat16 in training